_drafts/writing-a-bootloader-from-scratch.md

```diff
@@ -0,0 +1,449 @@
+
+Consider the following memory map:
+```
+/* memory_map.ld */
+MEMORY
+{
+  bootrom  (rx)  : ORIGIN = 0x00000000, LENGTH = 0x00010000
+  approm   (rx)  : ORIGIN = 0x00010000, LENGTH = 0x00004000
+  ram      (rwx) : ORIGIN = 0x20000000, LENGTH = 0x00004000
+  eram     (rwx) : ORIGIN = 0x20004000, LENGTH = 0x00004000
+}
+
+__bootrom_start__ = ORIGIN(bootrom);
+__bootrom_size__ = LENGTH(bootrom);
+__approm_start__ = ORIGIN(approm);
+__approm_size__ = LENGTH(approm);
+__eram_start__ = ORIGIN(eram);
+__eram_size__ = LENGTH(eram);
+```
+
+In this case, `approm` is our app storage and `eram` is our executable RAM, where we want to copy our program. Our bootloader needs to copy the cope from `approm` to `eram` before executing it.
+
+We know from our previous blog post that executable code typically ends up in the `.text` section we must tell the linker that this section is **stored** in `approm` but executed from `eram` so our program can execute correctly.
+
+This is similar to our `.data` section, which is stored in `rom` but lives in `ram` while the program is running. We use the `AT` linker command to specify the storage region and the `>` operator to specify the load region.  This is the resulting linker script section:
+
+```
+    .text :
+    {
+        KEEP(*(.vectors .vectors.*))
+        *(.text*)
+        *(.rodata*)
+    } > eram AT > approm
+```
+
+We then update our bootloader to copy our code from one to the other before starting the app:
+
+```c 
+  /* booloader.c */
+
+  /* copy app code to eram */
+  uint32_t *src = (uint32_t*) &__approm_start__;
+  uint32_t *dst = (uint32_t*) &__eram_start__;
+  int size = (int) &__approm_size__;
+  printf("Copying firmware from %p to %p\n", src, dst);
+  memcpy((uint8_t*) dst, (uint8_t*) src, size);
+
+  /* find app start & SP */
+  uint32_t app_sp = dst[0];
+  uint32_t app_start = dst[0];
+
+  /* cleanup peripherals here we may have initialized */
+
+  /* start the app */
+  start_app(app_start, app_sp);
+```
+
+
+### Locking the bootloader
+
+Last but not least, we can protect the bootloader using the memory protection unit to make it inaccessible from the app. This prevents accidentally erasing the bootloader during execution.
+
+If do not know about the MPU, check out Chris’s excellent blog post from a few weeks ago.
+
+Remember that our MPU regions must be power-of-2 sized. Thankfully, our bootloader already is! `0x4000` is 2^14 bytes.
+
+We add the following MPU code to our bootloader:
+
+```c
+int main(void) {
+  /* ... */
+  base_addr = 0x0;
+  *mpu_rbar = (base_addr | 1 << 4 | 1);
+  //  AP=0b110 to make the region read-only regardless of privilege
+  //  TEXSCB=0b000010 because the Code is in "Flash memory"
+  //  SIZE=13 because we want to cover 16kiB
+  //  ENABLE=1
+  *mpu_rasr = (0b110 << 24) | (0b000010 << 16) | (13 << 1) | 0x1;
+
+  start_app(app_start, app_sp);
+
+  /* Not reached */
+  while (1) {}
+}
+```
```

Comment:
Maybe add some sort of closing and the usual 

"See anything you'd like to change? Submit a pull request on [Github](https://github.com/memfault/interrupt)"

Reply:
Done